stdlib/benchmarks/collections/bench_dict.mojo

```diff
@@ -116,17 +86,43 @@ fn bench_dict_large_lookup(inout b: Bencher) raises:
 def main():
     seed()
     var m = Bench(BenchConfig(num_repetitions=1, warmup_iters=100))
-    m.bench_function[bench_dict_ctor](BenchId("bench_dict_ctor"))
-    m.bench_function[bench_dict_small_insert](
-        BenchId("bench_dict_small_insert")
-    )
-    m.bench_function[bench_dict_large_insert](
-        BenchId("bench_dict_large_insert")
-    )
-    m.bench_function[bench_dict_small_lookup](
-        BenchId("bench_dict_small_lookup")
-    )
-    m.bench_function[bench_dict_large_lookup](
-        BenchId("bench_dict_large_lookup")
+    m.bench_function[bench_dict_init](BenchId("bench_dict_init"))
+    alias sizes = (
+        10,
+        50,
+        100,
+        500,
+        1000,
+        5000,
+        10_000,
+        50_000,
+        100_000,
+        300_000,
+        400_000,
+        500_000,
+        600_000,
+        700_000,
+        800_000,
+        900_000,
+        1_000_000,
     )
+
+    @parameter
+    for i in range(len(sizes)):
+        alias size = sizes.get[i, Int]()
+        m.bench_function[bench_dict_insert[size]](
+            BenchId("bench_dict_insert[" + str(size) + "]")
+        )
+        m.bench_function[bench_dict_lookup[size]](
+            BenchId("bench_dict_lookup[" + str(size) + "]")
+        )
+
     m.dump_report()
+
+    @parameter
+    for i in range(len(sizes)):
+        alias size = sizes.get[i, Int]()
+        var mem_s = sizeof[Dict[Int, Int]]() * size
```

Comment:
Here we would actually need to know how many bytes are allocated in all internal fields (there could be pointers, non linearly growing data w/ regards to size, etc.). So we will likely need to add some kind of trait that returns the size of allocated bytes for a type and implement it for all types whose memory footprint is to be benchmarked.
Something like:
```mojo
trait MemoryMeasurable:
    fn _total_bytes_used(self):
        """Calculate total bytes used by the instance including `sizeof[Self]()`."""
        ...

struct Dict:
    ...
    fn _total_bytes_used(self):
        # subtract sizeof type because sizeof[Self]() already includes that (?)
        var entries = self._entries._total_bytes_used() - sizeof[V]
        var index = self._index._total_bytes_used() - sizeof[K]
        return sizeof[Self]() + entries + index
```

Reply:
Hello, very cool, here is a review for that part of the PR.

For the size that a Dict uses (in byte), you'd have to use a formula:

```mojo
alias elements = 1000
var tmp_reserved = upper_nearest_power_of_two(elements) #1024
#the resize up if overloaded
if (elements*3)>(tmp_reserved*2): tmp_reserved *= 2

var base = tmp_reserved

base *= sizeof[Optional[DictEntry[K,V]]]() #allocated self._entries
#self._entries and self._index structs + _reserved:Int, _n_entries: Int, size:Int
base += sizeof[Dict[K,V]]() 

#the allocated memory by index table:
if tmp_reserved <= 128:
    base+= sizeof[Scalar[DType.int8]]()*tmp_reserved
elif tmp_reserved <= 2**16 - 2:
    base+= sizeof[Scalar[DType.int16]]()*tmp_reserved
elif ... #see the DictIndex.__init__() for more elifs

print(base, " bytes")
```

Important note:
if `V` or `K` of `Dict[K,V]` is a type that allocate, it would not be reflected.
(a List for example, have a known inlined size, and heap allocations on the pointer)

As you pointed out,

maybe we need a new method+trait on some types that says how much memory an instance currently uses.
(`type.total_memory_used()` or something alike)

Using conditional conformance, we could assert the total memory usage of a type instance,
even if it is composed of types that allocates but conform to the trait.

It might not be a priority but an idea to keep in mind for the future.